src/mongocxx/test/collection.cpp

```diff
@@ -2808,4 +2812,145 @@ TEST_CASE("Ensure that the WriteConcernError 'errInfo' object is propagated", "[
+    };
+
+    // Set up our test environment:
+    instance::current();
+
+    auto client_opts = mongocxx::options::client();
+
+    // Listen to the insertion-failed event: we want to get a copy of the server's
+    // response so that we can compare it to the thrown exception later:
+    mongocxx::options::apm apm_opts;
+    apm_opts.on_command_failed([&writeErrors_well_formed](const events::command_failed_event& ev) {
+        writeErrors_well_formed(ev.failure());
+    });
+
+    apm_opts.on_command_succeeded(
+        [&writeErrors_well_formed](const mongocxx::events::command_succeeded_event& ev) {
+            writeErrors_well_formed(ev.reply());
+        });
+
+    client_opts.apm_opts(apm_opts);
+
+    auto mongodb_client = mongocxx::client(uri{}, client_opts);
+
+    database db = mongodb_client["prose_test_expose_details"];
+
+    const std::string collname{"mongo_cxx_driver-expose_details"};
+
+    // Drop the existing collection, if any:
+    db[collname].drop();
+
+    // Make a new collection with validation checking:
+    collection coll = db.create_collection(
+        collname,
+        make_document(kvp("validator",
+                          make_document(kvp("field_x", make_document(kvp("$type", "string")))))));
+
+    SECTION("insert a valid document") {
+        bsoncxx::builder::basic::document entry;
+
+        entry.append(kvp("_id", bsoncxx::oid()), kvp("field_x", "value0"));
+
+        REQUIRE_NOTHROW(coll.insert_one(entry.view()));
+    }
+
+    SECTION("cause a type violation on insert") {
+        bsoncxx::builder::basic::document entry;
+
+        entry.append(kvp("_id", bsoncxx::oid()), kvp("field_x", 42));
+
+        try {
+            coll.insert_one(entry.view());
+
+            // We should not make it here:
+            CHECK(false);
+        } catch (const operation_exception& e) {
+            auto rse = e.raw_server_error();
+
+            CHECK(rse.has_value());
+
+            CHECK(writeErrors_well_formed(*rse));
+        } catch (...) {
+            // An exception was thrown, but of the wrong type:
+            CHECK(false);
+        }
+    }
+
+    SECTION("can update valid object with a valid value type") {
+        auto oid_token = bsoncxx::oid();
+
+        // Make sure we have a valid entry to update:
+        bsoncxx::builder::basic::document original_entry;
+        original_entry.append(kvp("_id", oid_token), kvp("field_x", "value0"));
+        REQUIRE_NOTHROW(coll.insert_one(original_entry.view()));
+
+        // Update the same document:
+        bsoncxx::builder::basic::document filter, entry;
+
+        filter.append(kvp("_id", oid_token));
+        entry.append(kvp("$set", make_document(kvp("field_x", "value_updated"))));
+
+        REQUIRE_NOTHROW(coll.update_one(filter.view(), entry.view()));
+    }
+
+    SECTION("should NOT be able to update valid object with an invalid type") {
+        auto oid_token = bsoncxx::oid();
+
+        // Construct a valid entry to update:
```

Comment:
It might be nice to put this "construct valid entry" stuff into a function; I'm slightly on the fence at the moment. Currently would rather make sure the test details are right, but it wouldn't take much to convince me.

Reply:
I suggest removing this test. The only necessary test is "cause a type violation on insert".